File: PoeEye/PoeShared.Blazor.Controls/collection/ReactiveCollectionPresenter.cs
using System;
using System.Collections.Generic;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using PoeShared.Blazor;
using PoeShared.Blazor.Controls.Services;
using PoeShared.Blazor.Scaffolding;
using PoeShared.Scaffolding;

namespace PoeShared.Blazor.Controls;

public partial class ReactiveCollectionPresenter<TItem, TKey> : BlazorReactiveComponent
    where TItem : notnull
    where TKey : notnull
{
    public static readonly string JsFilePath = "./_content/PoeShared.Blazor.Controls/js/ReactiveCollectionPresenter.js";

    private readonly SemaphoreSlim applyGate = new(1, 1);
    private readonly SerialDisposable frameSubscription = new();
    private readonly TaskCompletionSource<bool> sessionReadySource = new(TaskCreationOptions.RunContinuationsAsynchronously);
    private readonly JsonSerializerOptions keySerializerOptions = new(JsonSerializerDefaults.Web);

    private Lazy<Task<IJSObjectReference>>? moduleTask;
    private IJSObjectReference? sessionReference;
    private ElementReference emptyStateElementRef;
    private Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> itemsByTransportKey = new();
    private IObservable<ReactiveCollectionFrame<TItem, TKey>>? subscribedFrames;
    private TimeSpan subscribedBatchDelay;

    [Inject] internal IReactiveCollectionItemRegistry ItemRegistry { get; init; } = default!;

    [Parameter] public IObservable<ReactiveCollectionFrame<TItem, TKey>>? Frames { get; set; }

    [Parameter, EditorRequired] public RenderFragment<TItem>? ItemTemplate { get; set; }

    /// <summary>
    /// Optional shell content shown when there are no visible items in the presented collection.
    /// This is owned by the presenter because emptiness is collection state; callers should not
    /// need an extra reactive wrapper just to toggle an empty-state message.
    /// </summary>
    [Parameter] public RenderFragment? EmptyTemplate { get; set; }

    [Parameter] public TimeSpan BatchDelay { get; set; } = TimeSpan.Zero;

    /// <summary>
    /// Advanced entry point: the host element is separate from the rendered island and can be styled independently.
    /// The tag is stable for the lifetime of an item. Changing it requires a remove/add and is intentionally not supported by update ops.
    /// </summary>
    [Parameter] public string ItemTagName { get; set; } = "div";

    [Parameter] public string? ItemClass { get; set; }

    [Parameter] public string? ItemStyle { get; set; }

    [Parameter] public Func<TItem, string?>? ItemClassSelector { get; set; }

    [Parameter] public Func<TItem, string?>? ItemStyleSelector { get; set; }

    protected override void OnInitialized()
    {
        base.OnInitialized();
    }

    protected override void OnParametersSet()
    {
        base.OnParametersSet();
        UpdateFrameSubscription();
    }

    private readonly AtomicFlag isRendered = new AtomicFlag();

    protected override bool ShouldRender()
    {
        return isRendered.Set();
    }

    protected override void OnAfterRender(bool firstRender)
    {
        base.OnAfterRender(firstRender);
    }

    protected override async Task OnAfterFirstRenderAsync()
    {
        await base.OnAfterFirstRenderAsync();

        moduleTask ??= new Lazy<Task<IJSObjectReference>>(() =>
            JsRuntime.InvokeAsync<IJSObjectReference>("import", JsFilePath).AsTask());

        try
        {
            var module = await moduleTask.Value;
            object? emptyStateElement = EmptyTemplate == null ? null : emptyStateElementRef;
            sessionReference = await module.InvokeAsync<IJSObjectReference>(
                "createSession",
                ElementRef,
                emptyStateElement,
                ReactiveCollectionItemHost.ComponentIdentifier);
            sessionReadySource.TrySetResult(true);
        }
        catch (Exception e)
        {
            sessionReadySource.TrySetException(e);
            throw;
        }
    }

    public override async ValueTask DisposeAsync()
    {
        try
        {
            sessionReadySource.TrySetCanceled();
            frameSubscription.Dispose();

            if (sessionReference != null)
            {
                await sessionReference.InvokeVoidAsync("dispose");
                await sessionReference.DisposeJsSafeAsync();
                sessionReference = null;
            }

            if (moduleTask is { IsValueCreated: true })
            {
                var module = await moduleTask.Value;
                await module.DisposeJsSafeAsync();
            }
        }
        catch (Exception e) when (e is OperationCanceledException or ObjectDisposedException or InvalidOperationException)
        {
            // During shutdown/reload the circuit may already be gone.
        }
        finally
        {
            itemsByTransportKey.Clear();
            applyGate.Dispose();
        }

        await base.DisposeAsync();
    }

    private void UpdateFrameSubscription()
    {
        if (ReferenceEquals(subscribedFrames, Frames) && subscribedBatchDelay == BatchDelay)
        {
            return;
        }

        subscribedFrames = Frames;
        subscribedBatchDelay = BatchDelay;

        // This presenter is intentionally render-once after initialization, so the frame pipeline
        // must be updated explicitly from parameters rather than relying on incidental rerenders.
        frameSubscription.Disposable = CreateFrameBatchStream(Frames, BatchDelay)
            .Where(x => x.Count > 0)
            .SubscribeAsync(ApplyFrameBatchAsync);
    }

    private IObservable<IReadOnlyList<ReactiveCollectionFrame<TItem, TKey>>> CreateFrameBatchStream(
        IObservable<ReactiveCollectionFrame<TItem, TKey>>? frames,
        TimeSpan batchDelay)
    {
        if (frames == null)
        {
            return Observable.Empty<IReadOnlyList<ReactiveCollectionFrame<TItem, TKey>>>();
        }

        var nonEmptyFrames = frames.Where(x => x != null && x.HasOperations);
        return batchDelay > TimeSpan.Zero
            ? nonEmptyFrames.Buffer(batchDelay).Where(x => x.Count > 0).Select(x => (IReadOnlyList<ReactiveCollectionFrame<TItem, TKey>>)x)
            : nonEmptyFrames.Select(x => (IReadOnlyList<ReactiveCollectionFrame<TItem, TKey>>)new[] { x });
    }

    private async Task ApplyFrameBatchAsync(IReadOnlyList<ReactiveCollectionFrame<TItem, TKey>> frameBatch)
    {
        if (frameBatch.Count == 0)
        {
            return;
        }

        await applyGate.WaitAsync();
        try
        {
            await sessionReadySource.Task;

            if (sessionReference == null)
            {
                return;
            }

            var workingState = new Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>>(itemsByTransportKey);
            var jsFrames = new List<ReactiveCollectionJsFrame>(frameBatch.Count);

            foreach (var frame in frameBatch)
            {
                var jsOperations = new List<ReactiveCollectionJsOperation>(frame.Operations.Count);
                foreach (var operation in frame.Operations)
                {
                    TranslateOperation(operation, workingState, jsOperations);
                }

                if (jsOperations.Count > 0)
                {
                    jsFrames.Add(new ReactiveCollectionJsFrame
                    {
                        Operations = jsOperations
                    });
                }
            }

            if (jsFrames.Count == 0)
            {
                itemsByTransportKey = workingState;
                return;
            }

            await sessionReference.InvokeVoidAsync("applyFrames", jsFrames);
            itemsByTransportKey = workingState;
        }
        finally
        {
            applyGate.Release();
        }
    }

    private void TranslateOperation(
        ReactiveCollectionOperation<TItem, TKey> operation,
        Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> workingState,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        switch (operation)
        {
            case ReactiveCollectionAddOperation<TItem, TKey> addOperation:
                TranslateAdd(addOperation, workingState, jsOperations);
                break;
            case ReactiveCollectionUpdateOperation<TItem, TKey> updateOperation:
                TranslateUpdate(updateOperation, workingState, jsOperations);
                break;
            case ReactiveCollectionMoveOperation<TItem, TKey> moveOperation:
                TranslateMove(moveOperation, jsOperations);
                break;
            case ReactiveCollectionRemoveOperation<TItem, TKey> removeOperation:
                TranslateRemove(removeOperation, workingState, jsOperations);
                break;
            case ReactiveCollectionClearOperation<TItem, TKey>:
                workingState.Clear();
                jsOperations.Add(new ReactiveCollectionJsOperation
                {
                    Kind = ReactiveCollectionJsOperationKind.Clear
                });
                break;
            case ReactiveCollectionResetOperation<TItem, TKey> resetOperation:
                TranslateReset(resetOperation, workingState, jsOperations);
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(operation), operation, $"Unsupported collection operation {operation.GetType().Name}.");
        }
    }

    private void TranslateAdd(
        ReactiveCollectionAddOperation<TItem, TKey> operation,
        Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> workingState,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        var transportKey = FormatKey(operation.Key);
        var beforeTransportKey = operation.BeforeKey == null ? null : FormatKey(operation.BeforeKey);
        var registration = ItemRegistry.Register(CreateItemContent(operation.Item));
        var hostMetadata = CreateHostMetadata(operation.Item);

        workingState[transportKey] = new ReactiveCollectionPresenterItemState<TItem, TKey>(
            operation.Key,
            operation.Item,
            registration.RegistrationId,
            registration.Version,
            hostMetadata.TagName,
            hostMetadata.ClassName,
            hostMetadata.Style);

        jsOperations.Add(new ReactiveCollectionJsOperation
        {
            Kind = ReactiveCollectionJsOperationKind.Add,
            Key = transportKey,
            BeforeKey = beforeTransportKey,
            RegistrationId = registration.RegistrationId,
            RegistrationVersion = registration.Version,
            TagName = hostMetadata.TagName,
            ClassName = hostMetadata.ClassName,
            Style = hostMetadata.Style
        });
    }

    private void TranslateUpdate(
        ReactiveCollectionUpdateOperation<TItem, TKey> operation,
        Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> workingState,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        var transportKey = FormatKey(operation.Key);
        if (!workingState.TryGetValue(transportKey, out var existingItemState))
        {
            // Advanced callers may emit an update before the initial reset/add.
            // Converting that case to an add keeps the presenter usable without silently recreating existing roots.
            TranslateAdd(new ReactiveCollectionAddOperation<TItem, TKey>(operation.Key, operation.Item), workingState, jsOperations);
            return;
        }

        var registration = ItemRegistry.Update(existingItemState.RegistrationId, CreateItemContent(operation.Item));
        var hostMetadata = CreateHostMetadata(operation.Item);

        // The host tag is intentionally stable for the lifetime of an item.
        // If a selector starts returning a different tag for an existing item, we keep the original shell
        // so update and move operations can preserve the mounted Blazor island instance.
        workingState[transportKey] = existingItemState with
        {
            Item = operation.Item,
            RegistrationVersion = registration.Version,
            ClassName = hostMetadata.ClassName,
            Style = hostMetadata.Style
        };

        jsOperations.Add(new ReactiveCollectionJsOperation
        {
            Kind = ReactiveCollectionJsOperationKind.Update,
            Key = transportKey,
            RegistrationId = registration.RegistrationId,
            RegistrationVersion = registration.Version,
            ClassName = hostMetadata.ClassName,
            Style = hostMetadata.Style
        });
    }

    private void TranslateMove(
        ReactiveCollectionMoveOperation<TItem, TKey> operation,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        jsOperations.Add(new ReactiveCollectionJsOperation
        {
            Kind = ReactiveCollectionJsOperationKind.Move,
            Key = FormatKey(operation.Key),
            BeforeKey = operation.BeforeKey == null ? null : FormatKey(operation.BeforeKey)
        });
    }

    private void TranslateRemove(
        ReactiveCollectionRemoveOperation<TItem, TKey> operation,
        Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> workingState,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        var transportKey = FormatKey(operation.Key);
        workingState.Remove(transportKey);
        jsOperations.Add(new ReactiveCollectionJsOperation
        {
            Kind = ReactiveCollectionJsOperationKind.Remove,
            Key = transportKey
        });
    }

    private void TranslateReset(
        ReactiveCollectionResetOperation<TItem, TKey> operation,
        Dictionary<string, ReactiveCollectionPresenterItemState<TItem, TKey>> workingState,
        List<ReactiveCollectionJsOperation> jsOperations)
    {
        workingState.Clear();

        var resetItems = new List<ReactiveCollectionJsItem>(operation.Items.Count);
        foreach (var itemSnapshot in operation.Items)
        {
            var transportKey = FormatKey(itemSnapshot.Key);
            var registration = ItemRegistry.Register(CreateItemContent(itemSnapshot.Item));
            var hostMetadata = CreateHostMetadata(itemSnapshot.Item);

            workingState[transportKey] = new ReactiveCollectionPresenterItemState<TItem, TKey>(
                itemSnapshot.Key,
                itemSnapshot.Item,
                registration.RegistrationId,
                registration.Version,
                hostMetadata.TagName,
                hostMetadata.ClassName,
                hostMetadata.Style);

            resetItems.Add(new ReactiveCollectionJsItem
            {
                Key = transportKey,
                RegistrationId = registration.RegistrationId,
                RegistrationVersion = registration.Version,
                TagName = hostMetadata.TagName,
                ClassName = hostMetadata.ClassName,
                Style = hostMetadata.Style
            });
        }

        jsOperations.Add(new ReactiveCollectionJsOperation
        {
            Kind = ReactiveCollectionJsOperationKind.Reset,
            Items = resetItems
        });
    }

    private RenderFragment CreateItemContent(TItem item)
    {
        if (ItemTemplate == null)
        {
            throw new InvalidOperationException($"{GetType().Name} requires {nameof(ItemTemplate)} to be set.");
        }

        return ItemTemplate(item);
    }

    private ReactiveCollectionHostMetadata CreateHostMetadata(TItem item)
    {
        var resolvedTagName = string.IsNullOrWhiteSpace(ItemTagName) ? "div" : ItemTagName;
        var resolvedClassName = ItemClassSelector?.Invoke(item) ?? ItemClass;
        var resolvedStyle = ItemStyleSelector?.Invoke(item) ?? ItemStyle;
        return new ReactiveCollectionHostMetadata(resolvedTagName, resolvedClassName, resolvedStyle);
    }

    private string FormatKey(TKey key)
    {
        return JsonSerializer.Serialize(key, keySerializerOptions);
    }

    private sealed record ReactiveCollectionPresenterItemState<TValue, TValueKey>(
        TValueKey Key,
        TValue Item,
        string RegistrationId,
        long RegistrationVersion,
        string TagName,
        string? ClassName,
        string? Style)
        where TValueKey : notnull;

    private readonly record struct ReactiveCollectionHostMetadata(string TagName, string? ClassName, string? Style);

    private sealed class ReactiveCollectionJsFrame
    {
        [JsonPropertyName("operations")]
        public required List<ReactiveCollectionJsOperation> Operations { get; init; }
    }

    private sealed class ReactiveCollectionJsItem
    {
        [JsonPropertyName("key")]
        public required string Key { get; init; }

        [JsonPropertyName("registrationId")]
        public required string RegistrationId { get; init; }

        [JsonPropertyName("registrationVersion")]
        public required long RegistrationVersion { get; init; }

        [JsonPropertyName("tagName")]
        public required string TagName { get; init; }

        [JsonPropertyName("className")]
        public string? ClassName { get; init; }

        [JsonPropertyName("style")]
        public string? Style { get; init; }
    }

    private sealed class ReactiveCollectionJsOperation
    {
        [JsonPropertyName("kind")]
        public required string Kind { get; init; }

        [JsonPropertyName("key")]
        public string? Key { get; init; }

        [JsonPropertyName("beforeKey")]
        public string? BeforeKey { get; init; }

        [JsonPropertyName("registrationId")]
        public string? RegistrationId { get; init; }

        [JsonPropertyName("registrationVersion")]
        public long? RegistrationVersion { get; init; }

        [JsonPropertyName("tagName")]
        public string? TagName { get; init; }

        [JsonPropertyName("className")]
        public string? ClassName { get; init; }

        [JsonPropertyName("style")]
        public string? Style { get; init; }

        [JsonPropertyName("items")]
        public List<ReactiveCollectionJsItem>? Items { get; init; }
    }

    private static class ReactiveCollectionJsOperationKind
    {
        public const string Add = "add";
        public const string Update = "update";
        public const string Move = "move";
        public const string Remove = "remove";
        public const string Clear = "clear";
        public const string Reset = "reset";
    }
}

File: PoeEye/PoeShared.Blazor.Controls/wwwroot/ts/ReactiveCollectionPresenter.ts
type RootComponentParameters = Record<string, unknown>;

interface IDynamicRootComponent {
    setParameters(parameters: RootComponentParameters): void;
    dispose(): Promise<void>;
}

interface IRootComponentsFunctions {
    add(toElement: Element, componentIdentifier: string, initialParameters: RootComponentParameters): Promise<IDynamicRootComponent>;
}

interface IBlazor {
    rootComponents: IRootComponentsFunctions;
}

interface ReactiveCollectionJsItem {
    key: string;
    registrationId: string;
    registrationVersion: number;
    tagName: string;
    className?: string | null;
    style?: string | null;
}

interface ReactiveCollectionJsOperation extends Partial<ReactiveCollectionJsItem> {
    kind: "add" | "update" | "move" | "remove" | "clear" | "reset";
    beforeKey?: string | null;
    items?: ReactiveCollectionJsItem[];
}

interface ReactiveCollectionJsFrame {
    operations: ReactiveCollectionJsOperation[];
}

interface ReactiveCollectionItemHandle {
    hostElement: HTMLElement;
    rootComponent: IDynamicRootComponent;
    tagName: string;
}

const BASE_ITEM_CLASS = "poe-reactive-collection-item";
const BlazorFacade: IBlazor | undefined = (window as any).Blazor;

if (!BlazorFacade) {
    throw new Error("Blazor root components API is not available on window.Blazor.");
}

class ReactiveCollectionSession {
    private readonly itemsByKey = new Map<string, ReactiveCollectionItemHandle>();
    private pendingWork: Promise<void> = Promise.resolve();

    constructor(
        private readonly container: HTMLElement,
        private readonly emptyStateElement: HTMLElement | null,
        private readonly componentIdentifier: string
    ) {
        this.updateEmptyState();
    }

    public applyFrames(frames: ReactiveCollectionJsFrame[]): Promise<void> {
        this.pendingWork = this.pendingWork.then(async () => {
            for (const frame of frames) {
                await this.applyFrame(frame);
            }
        });

        return this.pendingWork;
    }

    public async dispose(): Promise<void> {
        await this.pendingWork;
        await this.clearItems();
    }

    private async applyFrame(frame: ReactiveCollectionJsFrame): Promise<void> {
        for (const operation of frame.operations ?? []) {
            switch (operation.kind) {
                case "add":
                    await this.addItem(operation);
                    break;
                case "update":
                    this.updateItem(operation);
                    break;
                case "move":
                    this.moveItem(operation.key, operation.beforeKey);
                    break;
                case "remove":
                    await this.removeItem(operation.key);
                    break;
                case "clear":
                    await this.clearItems();
                    break;
                case "reset":
                    await this.resetItems(operation.items ?? []);
                    break;
                default:
                    throw new Error(`Unsupported collection operation '${operation.kind}'.`);
            }
        }

        this.updateEmptyState();
    }

    private async resetItems(items: ReactiveCollectionJsItem[]): Promise<void> {
        await this.clearItems();

        if (!items || items.length === 0) {
            return;
        }

        const fragment = document.createDocumentFragment();
        const pendingRoots: Array<Promise<void>> = [];

        for (const item of items) {
            const hostElement = this.createHostElement(item);
            fragment.appendChild(hostElement);
            pendingRoots.push(this.mountItem(item, hostElement));
        }

        this.container.appendChild(fragment);
        await Promise.all(pendingRoots);
    }

    private async addItem(operation: ReactiveCollectionJsOperation): Promise<void> {
        if (!operation.key || !operation.registrationId || operation.registrationVersion === undefined || !operation.tagName) {
            throw new Error("Add operation is missing required fields.");
        }

        const item: ReactiveCollectionJsItem = {
            key: operation.key,
            registrationId: operation.registrationId,
            registrationVersion: operation.registrationVersion,
            tagName: operation.tagName,
            className: operation.className,
            style: operation.style
        };

        const hostElement = this.createHostElement(item);
        this.insertHostElement(hostElement, operation.beforeKey);
        await this.mountItem(item, hostElement);
    }

    private updateItem(operation: ReactiveCollectionJsOperation): void {
        if (!operation.key) {
            throw new Error("Update operation is missing a key.");
        }

        const existingHandle = this.itemsByKey.get(operation.key);
        if (!existingHandle) {
            return;
        }

        this.applyHostState(existingHandle.hostElement, operation.className, operation.style);

        if (operation.registrationId && operation.registrationVersion !== undefined) {
            existingHandle.rootComponent.setParameters({
                registrationId: operation.registrationId,
                version: operation.registrationVersion
            });
        }
    }

    private moveItem(key?: string | null, beforeKey?: string | null): void {
        if (!key) {
            throw new Error("Move operation is missing a key.");
        }

        const existingHandle = this.itemsByKey.get(key);
        if (!existingHandle) {
            return;
        }

        const anchorHandle = beforeKey ? this.itemsByKey.get(beforeKey) : undefined;
        const anchorElement = anchorHandle?.hostElement ?? null;

        if (anchorElement === existingHandle.hostElement) {
            return;
        }

        // Move is intentionally DOM-only: the existing host element and mounted root component
        // are preserved so callers can reorder hot collections without recreating Blazor islands.
        this.container.insertBefore(existingHandle.hostElement, anchorElement);
    }

    private async removeItem(key?: string | null): Promise<void> {
        if (!key) {
            throw new Error("Remove operation is missing a key.");
        }

        const existingHandle = this.itemsByKey.get(key);
        if (!existingHandle) {
            return;
        }

        this.itemsByKey.delete(key);
        await existingHandle.rootComponent.dispose();
        existingHandle.hostElement.remove();
    }

    private async clearItems(): Promise<void> {
        const handles = Array.from(this.itemsByKey.values());
        this.itemsByKey.clear();

        for (const handle of handles) {
            await handle.rootComponent.dispose();
            handle.hostElement.remove();
        }
    }

    private updateEmptyState(): void {
        if (!this.emptyStateElement) {
            return;
        }

        const hasItems = this.itemsByKey.size > 0;
        this.container.hidden = !hasItems;
        this.emptyStateElement.hidden = hasItems;
    }

    private createHostElement(item: ReactiveCollectionJsItem): HTMLElement {
        const tagName = item.tagName?.trim() || "div";
        const hostElement = document.createElement(tagName);
        hostElement.dataset["reactiveCollectionKey"] = item.key;
        this.applyHostState(hostElement, item.className, item.style);
        return hostElement;
    }

    private applyHostState(hostElement: HTMLElement, className?: string | null, style?: string | null): void {
        hostElement.className = [BASE_ITEM_CLASS, className ?? ""].filter(Boolean).join(" ");

        if (style && style.length > 0) {
            hostElement.setAttribute("style", style);
        } else {
            hostElement.removeAttribute("style");
        }
    }

    private insertHostElement(hostElement: HTMLElement, beforeKey?: string | null): void {
        const anchorHandle = beforeKey ? this.itemsByKey.get(beforeKey) : undefined;
        this.container.insertBefore(hostElement, anchorHandle?.hostElement ?? null);
    }

    private async mountItem(item: ReactiveCollectionJsItem, hostElement: HTMLElement): Promise<void> {
        const rootComponent = await BlazorFacade.rootComponents.add(hostElement, this.componentIdentifier, {
            registrationId: item.registrationId,
            version: item.registrationVersion
        });

        this.itemsByKey.set(item.key, {
            hostElement,
            rootComponent,
            tagName: item.tagName
        });
    }
}

export function createSession(
    container: HTMLElement,
    emptyStateElement: HTMLElement | null,
    componentIdentifier: string
): ReactiveCollectionSession {
    if (!container) {
        throw new Error("Container element must be specified.");
    }

    if (!componentIdentifier) {
        throw new Error("Component identifier must be specified.");
    }

    return new ReactiveCollectionSession(container, emptyStateElement, componentIdentifier);
}
